### PR TITLE
[CSPL-486] Update MC configMap for  "SPLUNK_MULTISITE_MASTER" and "SPLUNK_SITE" env var

### PR DESCRIPTION
**Pull request by kashok-splunk**
**Scenario:**
 - Multisite-IDXC is running.
 - Create a standalone instance, this changes the env variables on the MC pod and MC loses the multi-site info

**Solution:**
- Append "SPLUNK_MULTISITE_MASTER" and "SPLUNK_SITE" env for MC in MC configMap
- Set multiple owners for the stateful set

### CODE REVIEW

File: pkg/splunk/controller/statefulset_test.go [hunk 1]
```diff
@@ -52,3 +56,65 @@ func TestDefaultStatefulSetPodManager(t *testing.T) {
...
+	if err.Error() != "NotFound" {
+		t.Errorf("Couldn't detect missing owner reference %s", current.GetName())
+	}
+
+	// Create statefulset
+	err = splutil.CreateResource(c, &current)
+	if err != nil {
+		t.Errorf("Failed to create owner reference  %s", current.GetName())
+	}
+	//	current.OwnerReferences = nil
```
Comment (akondur) [inline pkg/splunk/controller/statefulset_test.go:87]: Nit: remove this line?

File: pkg/splunk/controller/statefulset_test.go [hunk 2]
```diff
@@ -52,3 +56,65 @@ func TestDefaultStatefulSetPodManager(t *testing.T) {
...
+
+	// Test existing owner reference
+	err = SetStatefulSetOwnerRef(c, &cr, namespacedName)
```
Comment (akondur) [inline pkg/splunk/controller/statefulset_test.go:90]: Nit: For a little bit more easy coverage, we could call SetStatefulSetOwnerRef again to cover the existing OR scenario :)

File: pkg/splunk/enterprise/monitoringconsole.go [hunk 3]
```diff
@@ -87,10 +79,18 @@ func ApplyMonitoringConsole(client splcommon.ControllerClient, cr splcommon.Meta
 		if err != nil {
 			return err
 		}
-		multiSite = clusterInfo.MultiSite
+		multiSite := clusterInfo.MultiSite
+		if multiSite == "true" {
```
Comment (akondur) [inline pkg/splunk/enterprise/monitoringconsole.go:83]: (if multSite) vs (if multiSite == "true")?

File: pkg/splunk/enterprise/monitoringconsole.go [hunk 4]
```diff
@@ -194,27 +193,20 @@ func getMonitoringConsoleStatefulSet(client splcommon.ControllerClient, cr splco
...
 	//update podTemplate annotation with configMap resource version
-	//use emptyEnv here to avoid adding site amd multisite_master info in configMap
-	//TODO: Update this configMap retrieval by API from PR #134
-	if cr.GetObjectMeta().GetDeletionTimestamp() != nil {
-		monitoringConsoleConfigMap, _ = ApplyMonitoringConsoleEnvConfigMap(client, cr.GetNamespace(), cr.GetName(), emptyEnv, false)
-	} else {
-		monitoringConsoleConfigMap, _ = ApplyMonitoringConsoleEnvConfigMap(client, cr.GetNamespace(), cr.GetName(), emptyEnv, true)
-	}
+	namespacedName := types.NamespacedName{Namespace: cr.GetNamespace(), Name: configMap}
+	monitoringConsoleConfigMap, _ = splctrl.GetConfigMap(client, namespacedName)
```
Comment (akondur) [inline pkg/splunk/enterprise/monitoringconsole.go:204]: Are we not checking for error here?

File: pkg/splunk/enterprise/monitoringconsole.go [hunk 5]
```diff
@@ -194,27 +193,20 @@ func getMonitoringConsoleStatefulSet(client splcommon.ControllerClient, cr splco
...
 	statefulSet.Spec.Template.ObjectMeta.Annotations[monitoringConsoleConfigRev] = monitoringConsoleConfigMap.ResourceVersion
 
-	statefulSet.SetOwnerReferences(append(statefulSet.GetOwnerReferences(), splcommon.AsOwner(cr, false)))
+	//set owner reference for splunk monitoring console statefulset
+	namespacedName = types.NamespacedName{Namespace: cr.GetNamespace(), Name: GetSplunkStatefulsetName(SplunkMonitoringConsole, cr.GetNamespace())}
+	_ = splctrl.SetStatefulSetOwnerRef(client, cr, namespacedName)
```
Comment (akondur) [inline pkg/splunk/enterprise/monitoringconsole.go:209]: Same comment about error handling as above.

**Review by akondur — CHANGES_REQUESTED** [review]

**Review by smohan-splunk — APPROVED** [review]
CSPL-498 & CSPL-499 identified and added to backlog as part of Code review

File: pkg/splunk/enterprise/monitoringconsole.go (hunk 3 above)
Comment (kashok-splunk) [reply pkg/splunk/enterprise/monitoringconsole.go:83]: "multSite" here is a string, not a bool

File: pkg/splunk/enterprise/monitoringconsole.go (hunk 4 above)
Comment (kashok-splunk) [reply pkg/splunk/enterprise/monitoringconsole.go]: Done

File: pkg/splunk/enterprise/monitoringconsole.go (hunk 5 above)
Comment (kashok-splunk) [reply pkg/splunk/enterprise/monitoringconsole.go]: Done

File: pkg/splunk/controller/statefulset_test.go [hunk 6]
```diff
@@ -52,3 +56,70 @@ func TestDefaultStatefulSetPodManager(t *testing.T) {
...
+	}
+
+	// Test existing owner reference
+	err = SetStatefulSetOwnerRef(c, &cr, namespacedName)
+	if err != nil {
+		t.Errorf("Couldn't set owner ref for owner reference %s", current.GetName())
+	}
+
+	// Try adding same owner again
+	err = SetStatefulSetOwnerRef(c, &cr, namespacedName)
```
Comment (akondur) [inline pkg/splunk/controller/statefulset_test.go:95]: Here, shouldn't we see an error?

File: pkg/splunk/enterprise/monitoringconsole.go (hunk 3 above)
Comment (akondur) [reply pkg/splunk/enterprise/monitoringconsole.go:83]: Ah ok my bad!

File: pkg/splunk/controller/statefulset_test.go [hunk 7]
```diff
@@ -52,3 +56,70 @@ func TestDefaultStatefulSetPodManager(t *testing.T) {
...
+	c := spltest.NewMockClient()
+
+	current := appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "splunk-test-monitoring-console",
+			Namespace: "test",
+		},
+	}
+
+	_, _ = ApplyStatefulSet(c, &current)
```
Comment (akondur) [inline pkg/splunk/controller/statefulset_test.go:118]: Error handle here?

File: pkg/splunk/enterprise/finalizers_test.go [hunk 8]
```diff
@@ -139,6 +146,14 @@ func splunkDeletionTester(t *testing.T, cr splcommon.MetaObject, delete func(spl
 
 	c := spltest.NewMockClient()
 	c.ListObj = &pvclist
+	statefulset := appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "splunk-test-monitoring-console",
+			Namespace: "test",
+		},
+	}
+	_, _ = splctrl.ApplyStatefulSet(c, &statefulset)
```
Comment (akondur) [inline pkg/splunk/enterprise/finalizers_test.go:155]: Error handle here?

File: pkg/splunk/enterprise/monitoringconsole_test.go [hunk 9]
```diff
@@ -57,28 +59,36 @@ func TestApplyMonitoringConsole(t *testing.T) {
...
 		{Name: "A", Value: "a"},
 	}
+
+	statefulset := appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "splunk-test-monitoring-console",
+			Namespace: "test",
+		},
+	}
+	_, _ = splctrl.ApplyStatefulSet(c, &statefulset)
```
Comment (akondur) [inline pkg/splunk/enterprise/monitoringconsole_test.go:84]: Error handle here?

**Review by akondur — APPROVED** [review]

File: pkg/splunk/controller/util.go [hunk 10]
```diff
@@ -140,15 +140,6 @@ func MergePodSpecUpdates(current *corev1.PodSpec, revised *corev1.PodSpec, name
 				result = true
 			}
 
-			// check Env
-			if splcommon.CompareEnvs(current.Containers[idx].Env, revised.Containers[idx].Env) {
```
Comment (gaurav-splunk) [inline pkg/splunk/controller/util.go:144]: basic question, why have we removed this?

File: pkg/splunk/controller/statefulset_test.go [hunk 11]
```diff
@@ -52,3 +56,70 @@ func TestDefaultStatefulSetPodManager(t *testing.T) {
...
+
+	namespacedName := types.NamespacedName{Namespace: "test", Name: "splunk-test-monitoring-console"}
+	_, err := GetstatefulSetByName(c, &cr, namespacedName)
```
Comment (gaurav-splunk) [inline pkg/splunk/controller/statefulset_test.go]: nit: %s/GetstatefulSetByName/GetStatefulSetByName

File: pkg/splunk/controller/statefulset_test.go [hunk 12]
```diff
@@ -52,3 +56,70 @@ func TestDefaultStatefulSetPodManager(t *testing.T) {
...
+	}
+
+	// Try adding same owner again
+	err = SetStatefulSetOwnerRef(c, &cr, namespacedName)
+	if err != nil {
+		t.Errorf("Couldn't set owner ref for owner reference %s", current.GetName())
+	}
+}
+
+func TestGetstatefulSetByName(t *testing.T) {
```
Comment (gaurav-splunk) [inline pkg/splunk/controller/statefulset_test.go:101]: nit : %s/TestGetstatefulSetByName/TestGetStatefulSetByName

File: pkg/splunk/controller/statefulset_test.go (hunk 6 above)
Comment (akondur) [reply pkg/splunk/controller/statefulset_test.go:95]: My bad! Error will be nil but we are not setting owner references again also :)

File: pkg/splunk/controller/statefulset_test.go (hunk 7 above)
Comment (kashok-splunk) [reply pkg/splunk/controller/statefulset_test.go]: Done

File: pkg/splunk/enterprise/monitoringconsole_test.go (hunk 9 above)
Comment (kashok-splunk) [reply pkg/splunk/enterprise/monitoringconsole_test.go]: Done

File: pkg/splunk/controller/statefulset_test.go [hunk 13]
```diff
@@ -52,3 +56,73 @@ func TestDefaultStatefulSetPodManager(t *testing.T) {
...
+	cr := enterprisev1.Standalone{
```
Comment (gaurav-splunk) [inline pkg/splunk/controller/statefulset_test.go:102]: Also, what is the need for this `cr` here? I know you are passing it in `GetstatefulSetByName`, but even in that function, you are not using it.

File: pkg/splunk/controller/statefulset.go [hunk 14]
```diff
@@ -230,3 +231,44 @@ func UpdateStatefulSetPods(c splcommon.ControllerClient, statefulSet *appsv1.Sta
...
+	err = splutil.UpdateResource(client, statefulset)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetstatefulSetByName retrieves current statefulset
+func GetstatefulSetByName(c splcommon.ControllerClient, cr splcommon.MetaObject, namespacedName types.NamespacedName) (*appsv1.StatefulSet, error) {
```
Comment (gaurav-splunk) [inline pkg/splunk/controller/statefulset.go:264]: I don't think we need `cr` here as we are not using it in this function.

**Review by gaurav-splunk — CHANGES_REQUESTED** [review]

File: pkg/splunk/controller/util.go (hunk 10 above)
Comment (kashok-splunk) [reply pkg/splunk/controller/util.go:144]: We don't need this check anymore, added when I was using configMap hash to update MC pos, now we use annotations

File: pkg/splunk/controller/statefulset.go (hunk 14 above)
Comment (kashok-splunk) [reply pkg/splunk/controller/statefulset.go:264]: my bad, I was playing around and forgot to change this. Let me fix this

File: pkg/splunk/controller/statefulset_test.go [hunk 15]
```diff
@@ -52,3 +56,66 @@ func TestDefaultStatefulSetPodManager(t *testing.T) {
...
+
+	err := SetStatefulSetOwnerRef(c, &cr, namespacedName)
+	if err.Error() != "NotFound" {
+		t.Errorf("Couldn't detect missing owner reference %s", current.GetName())
+	}
+
+	// Create statefulset
+	err = splutil.CreateResource(c, &current)
+	if err != nil {
+		t.Errorf("Failed to create owner reference  %s", current.GetName())
```
Comment (sgontla) [inline pkg/splunk/controller/statefulset_test.go:85]: We are actually trying to create the resource here?

**Review by gaurav-splunk — APPROVED** [review]

File: pkg/splunk/controller/statefulset_test.go [hunk 16]
```diff
@@ -52,3 +56,66 @@ func TestDefaultStatefulSetPodManager(t *testing.T) {
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "splunk-test-monitoring-console",
+			Namespace: "test",
+		},
+	}
+	namespacedName := types.NamespacedName{Namespace: "test", Name: "splunk-test-monitoring-console"}
+
+	err := SetStatefulSetOwnerRef(c, &cr, namespacedName)
+	if err.Error() != "NotFound" {
+		t.Errorf("Couldn't detect missing owner reference %s", current.GetName())
```
[the remark this answers is not in the document]
Comment (kashok-splunk) [reply pkg/splunk/controller/statefulset_test.go:79]: No, we create in line 83

File: pkg/splunk/controller/statefulset_test.go (hunk 15 above)
Comment (kashok-splunk) [reply pkg/splunk/controller/statefulset_test.go:85]: yes

File: pkg/splunk/controller/statefulset_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (sgontla) [reply pkg/splunk/controller/statefulset_test.go:79]: then, we need to correct the error right?

File: pkg/splunk/controller/statefulset_test.go (hunk 15 above)
Comment (sgontla) [reply pkg/splunk/controller/statefulset_test.go:85]: then, we need to correct the error right?

**Review by sgontla — APPROVED** [review]

**Pull request merged**